### PR TITLE
Json loader

### PR DESCRIPTION
**Pull request by jochman**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/45173

Changed all json imports to use the JSON_Handler.
It uses UJSON. We might want to change in the future to orjson or to json5 compatible parser.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -22,9 +22,9 @@ references:
       run:
         name: Create requirement files
         command: |
-          pip install pipfile-requirements
-          pipfile2req > requirements.txt
-          pipfile2req -d > requirements-dev.txt
+          pip install pipenv
```
Comment (dorschw) [inline .circleci/config.yml:25]: let's wait for https://github.com/demisto/demisto-sdk/pull/1963/ to be approved, then merge it into this one and proceed

Comment (dorschw) [inline .circleci/config.yml:25]: would it be better if we waited for https://github.com/demisto/demisto-sdk/pull/1963/ to be merged first?

File: demisto_sdk/commands/common/handlers/yaml/ruamel_handler.py [hunk 2]
```diff
@@ -26,7 +26,7 @@ def dump(self, data, stream, sort_keys=False):
             data = order_dict(data)
         self._yaml.dump(data, stream)
 
-    def dumps(self, data, sort_keys=False):
+    def dumps(self, data, sort_keys=False, indent=None):
```
Comment (dorschw) [inline demisto_sdk/commands/common/handlers/yaml/ruamel_handler.py:29]: if possible, please maintain a constant order of arguments (see `xsoar_handler.dumps` - there, `indent` comes before `sort_keys`)

File: demisto_sdk/commands/common/hook_validations/content_entity_validator.py [hunk 3]
```diff
@@ -23,6 +22,9 @@
                                                is_test_config_match,
                                                run_command)
 
+json = JSON_Handler()
+
+
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/content_entity_validator.py:27]:
```suggestion
```

File: demisto_sdk/commands/common/hook_validations/integration.py [hunk 4]
```diff
@@ -31,6 +30,9 @@
     get_file_version_suffix_if_exists, get_files_in_dir, get_item_marketplaces,
     get_pack_name, is_iron_bank_pack, print_error, server_version_compare)
 
+json = JSON_Handler()
+
+
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/integration.py:35]:
```suggestion
```

File: demisto_sdk/commands/common/hook_validations/structure.py [hunk 5]
```diff
@@ -17,14 +16,17 @@
...
 from demisto_sdk.commands.common.hook_validations.base_validator import \
     BaseValidator
 from demisto_sdk.commands.common.tools import (get_remote_file,
                                                is_file_path_in_pack)
 from demisto_sdk.commands.format.format_constants import \
     OLD_FILE_DEFAULT_1_FROMVERSION
 
+json = JSON_Handler()
+
+
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/structure.py:29]:
```suggestion
```

File: demisto_sdk/commands/download/downloader.py [hunk 6]
```diff
@@ -37,6 +36,9 @@
 from demisto_sdk.commands.format.format_module import format_manager
 from demisto_sdk.commands.split.ymlsplitter import YmlSplitter
 
+json = JSON_Handler()
+
+
```
Comment (dorschw) [inline demisto_sdk/commands/download/downloader.py:41]:
```suggestion
```

File: demisto_sdk/commands/format/update_generic_yml.py [hunk 7]
```diff
@@ -8,13 +7,16 @@
...
+from demisto_sdk.commands.common.handlers import JSON_Handler, YAML_Handler
 from demisto_sdk.commands.common.tools import (
     _get_file_id, find_type, get_entity_id_by_entity_type,
     get_not_registered_tests, get_scripts_and_commands_from_yml_data, get_yaml,
     is_uuid, listdir_fullpath)
 from demisto_sdk.commands.format.update_generic import BaseUpdate
 
+json = JSON_Handler()
+
+
```
Comment (dorschw) [inline demisto_sdk/commands/format/update_generic_yml.py:19]:
```suggestion
```

File: demisto_sdk/commands/generate_integration/code_generator.py [hunk 8]
```diff
@@ -19,6 +18,9 @@
 from demisto_sdk.commands.generate_integration.XSOARIntegration import \
     XSOARIntegration
 
+json = JSON_Handler()
+
+
```
Comment (dorschw) [inline demisto_sdk/commands/generate_integration/code_generator.py:23]:
```suggestion
```

File: demisto_sdk/commands/generate_integration/tests/code_generator_test.py [hunk 9]
```diff
@@ -1,13 +1,15 @@
...
-from demisto_sdk.commands.common.handlers import YAML_Handler
+from demisto_sdk.commands.common.handlers import JSON_Handler, YAML_Handler
 from demisto_sdk.commands.common.legacy_git_tools import git_path
 from demisto_sdk.commands.generate_integration.code_generator import (
     IntegrationGeneratorCommand, IntegrationGeneratorConfig,
     IntegrationGeneratorOutput, json_body_to_code)
 
+json = JSON_Handler()
+
+
```
Comment (dorschw) [inline demisto_sdk/commands/generate_integration/tests/code_generator_test.py:12]:
```suggestion
```

File: demisto_sdk/commands/generate_outputs/json_to_outputs/json_to_outputs.py [hunk 10]
```diff
@@ -71,17 +71,19 @@
...
 import dateparser
 
-from demisto_sdk.commands.common.handlers import YAML_Handler
+from demisto_sdk.commands.common.handlers import JSON_Handler, YAML_Handler
 from demisto_sdk.commands.common.tools import (LOG_COLORS, print_color,
                                                print_error)
 
+json = JSON_Handler()
+
+
```
Comment (dorschw) [inline demisto_sdk/commands/generate_outputs/json_to_outputs/json_to_outputs.py:86]:
```suggestion
```

File: demisto_sdk/commands/generate_test_playbook/test_playbook_generator.py [hunk 11]
```diff
@@ -1,13 +1,15 @@
...
 import click
 
 from demisto_sdk.commands.common.constants import FileType
-from demisto_sdk.commands.common.handlers import YAML_Handler
+from demisto_sdk.commands.common.handlers import JSON_Handler, YAML_Handler
 from demisto_sdk.commands.upload.uploader import Uploader
 
+json = JSON_Handler()
+
+
```
Comment (dorschw) [inline demisto_sdk/commands/generate_test_playbook/test_playbook_generator.py:12]:
```suggestion
```

File: demisto_sdk/commands/init/initiator.py [hunk 12]
```diff
@@ -20,13 +19,16 @@
...
+from demisto_sdk.commands.common.handlers import JSON_Handler, YAML_Handler
 from demisto_sdk.commands.common.tools import (LOG_COLORS,
                                                get_common_server_path,
                                                get_pack_name, print_error,
                                                print_v, print_warning)
 from demisto_sdk.commands.secrets.secrets import SecretsValidator
 
+json = JSON_Handler()
+
+
```
Comment (dorschw) [inline demisto_sdk/commands/init/initiator.py:31]:
```suggestion
```

File: demisto_sdk/commands/init/templates/BaseScript/BaseScript_test.py [hunk 13]
```diff
@@ -9,7 +9,8 @@
 """
 
 import io
-import json
+from demisto_sdk.commands.common.handlers import JSON_Handler
```
Comment (dorschw) [inline demisto_sdk/commands/init/templates/BaseScript/BaseScript_test.py:12]: perhaps it'd be better import it at the top of the file, to avoid mistaken uses of other json implementations elsewhere in the file

File: demisto_sdk/commands/init/tests/initiator_test.py [hunk 14]
```diff
@@ -12,10 +11,13 @@
...
     PACK_INITIAL_VERSION, PACK_SUPPORT_OPTIONS, XSOAR_AUTHOR, XSOAR_SUPPORT,
     XSOAR_SUPPORT_URL)
-from demisto_sdk.commands.common.handlers import YAML_Handler
+from demisto_sdk.commands.common.handlers import JSON_Handler, YAML_Handler
 from demisto_sdk.commands.init.initiator import Initiator
 from TestSuite.test_tools import ChangeCWD
 
+json = JSON_Handler()
+
+
```
Comment (dorschw) [inline demisto_sdk/commands/init/tests/initiator_test.py:20]:
```suggestion
```

File: demisto_sdk/commands/unify/tests/yml_unifier_test.py [hunk 15]
```diff
@@ -11,12 +10,15 @@
...
-from demisto_sdk.commands.common.handlers import YAML_Handler
+from demisto_sdk.commands.common.handlers import JSON_Handler, YAML_Handler
 from demisto_sdk.commands.common.legacy_git_tools import git_path
 from demisto_sdk.commands.common.tools import get_yaml
 from demisto_sdk.commands.unify.yml_unifier import YmlUnifier
 from TestSuite.test_tools import ChangeCWD
 
+json = JSON_Handler()
+
+
```
Comment (dorschw) [inline demisto_sdk/commands/unify/tests/yml_unifier_test.py:21]:
```suggestion
```

File: demisto_sdk/commands/update_release_notes/tests/update_rn_test.py [hunk 16]
```diff
@@ -664,16 +666,16 @@ class TestRNUpdateUnit:
...
-#### Incident Fields
+# Incident Fields
 - **XDR Alerts**
 
-#### Object Types
+# Object Types
 - **(Asset) - Sample GenericType**
 
-#### Object Fields
+# Object Fields
```
Comment (dorschw) [inline demisto_sdk/commands/update_release_notes/tests/update_rn_test.py:678]: is this change intentional?

File: demisto_sdk/commands/update_release_notes/tests/update_rn_test.py [hunk 17]
```diff
@@ -689,24 +691,24 @@ class TestRNUpdateUnit:
...
 - **(Asset) - Sample GenericField**
 
-#### Integrations
-##### Cortex XDR - IR
+# Integrations
+# Cortex XDR - IR
 - %%UPDATE_RN%%
 
-##### Sample
+# Sample
```
Comment (dorschw) [inline demisto_sdk/commands/update_release_notes/tests/update_rn_test.py:711]: intentional?

File: demisto_sdk/tests/integration_tests/format_integration_test.py [hunk 18]
```diff
@@ -29,6 +28,9 @@
     GENERIC_DEFINITION, GENERIC_FIELD, GENERIC_MODULE, GENERIC_TYPE)
 from TestSuite.test_tools import ChangeCWD
 
+json = JSON_Handler()
+
+
```
Comment (dorschw) [inline demisto_sdk/tests/integration_tests/format_integration_test.py:33]:
```suggestion
```

**Review by dorschw — COMMENTED** [review]
Awesome job!

File: demisto_sdk/commands/update_release_notes/tests/update_rn_test.py (hunk 17 above)
Comment (jochman) [reply demisto_sdk/commands/update_release_notes/tests/update_rn_test.py]: must be automatically 🤷

File: demisto_sdk/commands/init/templates/BaseScript/BaseScript_test.py (hunk 13 above)
Comment (jochman) [reply demisto_sdk/commands/init/templates/BaseScript/BaseScript_test.py:12]: isort!

File: demisto_sdk/tests/test_files/fake_integration_expected_test_playbook__all_brands.yml [hunk 19]
```diff
@@ -19,7 +19,7 @@ tasks:
       '#none#':
       - '1'
     separatecontext: false
-    view: '{"position": {"x": 50, "y": 50}}'
+    view: '{"position":{"x":50,"y":50}}'
```
Comment (dorschw) [inline demisto_sdk/tests/test_files/fake_integration_expected_test_playbook__all_brands.yml:22]: why are all these changes here? auto-formatted?

**Review by dorschw — CHANGES_REQUESTED** [review]
awesome job, as always, please check the comments.

File: demisto_sdk/commands/common/handlers/json/ujson_handler.py [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+        except ValueError as e:
+            raise JSONDecodeError(e)
+
+    def load(self, fp: TextIO):
+        try:
+            return ujson.load(fp)
+        except ValueError as e:
+            raise JSONDecodeError(e)
+
+    def dump(self, obj, fp: TextIO, sort_keys=False, indent=0):
```
[the remark this answers is not in the document]
Comment (jochman) [reply demisto_sdk/commands/common/handlers/json/ujson_handler.py:34]: if you're suppling a default - the typed are automatically determined.

File: demisto_sdk/tests/test_files/fake_integration_expected_test_playbook__all_brands.yml (hunk 19 above)
Comment (jochman) [reply demisto_sdk/tests/test_files/fake_integration_expected_test_playbook__all_brands.yml:22]: Its a kind of a problem in the unittest that I have ugly-solved.

**Review by dorschw — CHANGES_REQUESTED** [review]
Please address the unresolved issues (order of arguments)

**Review by dorschw — APPROVED** [review]

**Comment by idovandijk** [comment]
@jochman this is a pretty big change, can you test this on a couple of JSON objects - incident field, layout, etc? Meaning export from XSOAR (not SDK download), format, validate...
I will try to play with it as well but won't have a lot of capacity for this this week
